False positive in unneeded_synthesized_initializer

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

`unneeded_synthesized_initializer` reports a false positive when the init has side effects. In the attached sample, the `init` prints something, and removing it would remove the print. In the example where I encountered it in a real app, a SwiftUI view modifier's `init` contained a call to `UIScrollView.appearance().bounces = false`, which we definitely don't want to lose by deleting the `init`!

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Linting Swift files in current working directory
Linting 'InitTest.swift' (1/1)
/Users/igeek/Downloads/InitTest/InitTest.swift:2:5: warning: Unneeded Synthesized Initializer Violation: This default initializer would be synthesized automatically - you do not need to define it (unneeded_synthesized_initializer)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version: **0.52.3**
* Installation method used: **homebrew**
* Paste your configuration file: **n/a (default configuration)**
* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? **No**
* Which Xcode version are you using? **Xcode 14.3.1 (14E300c) on macOS 13.4.1 (22F82)**
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
struct MyStruct {
    init() {
        print("perform side effect")
    }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,7 +14,10 @@
 
 #### Bug Fixes
 
-* None.
+* Fix false positives for the `unneeded_synthesized_initializer` rule, when
+  no argument initializers had side-effects.  
+  [Martin Redington](https://github.com/mildm8nnered)
+  [#5075](https://github.com/realm/SwiftLint/issues/5075)
 
 ## 0.52.3: Duplicate Hampers
 

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift
@@ -222,7 +222,6 @@ private extension StructDeclSyntax {
         return initializerBody.statements.isEmpty
     }
 
-
     // Does the actual access level of an initializer match the access level of the synthesized
     // memberwise initializer?
     private func initializerModifiers(

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift
@@ -120,7 +120,8 @@ private extension StructDeclSyntax {
 
         return initializers.filter {
             self.initializerParameters($0.parameterList, match: storedProperties) &&
-            (($0.parameterList.isEmpty && hasNoSideEffects($0.body)) || initializerBody($0.body, matches: storedProperties)) &&
+            (($0.parameterList.isEmpty && hasNoSideEffects($0.body)) ||
+             initializerBody($0.body, matches: storedProperties)) &&
             initializerModifiers($0.modifiers, match: storedProperties) && !$0.isInlinable
         }
     }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift
@@ -120,7 +120,7 @@ private extension StructDeclSyntax {
 
         return initializers.filter {
             self.initializerParameters($0.parameterList, match: storedProperties) &&
-            ($0.parameterList.isEmpty || initializerBody($0.body, matches: storedProperties)) &&
+            (($0.parameterList.isEmpty && hasNoSideEffects($0.body)) || initializerBody($0.body, matches: storedProperties)) &&
             initializerModifiers($0.modifiers, match: storedProperties) && !$0.isInlinable
         }
     }
@@ -214,6 +214,14 @@ private extension StructDeclSyntax {
         }
         return statements.isEmpty
     }
+
+    private func hasNoSideEffects(_ initializerBody: CodeBlockSyntax?) -> Bool {
+        guard let initializerBody else {
+            return true
+        }
+        return initializerBody.statements.isEmpty
+    }
+
 
     // Does the actual access level of an initializer match the access level of the synthesized
     // memberwise initializer?

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRuleExamples.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRuleExamples.swift
@@ -180,6 +180,23 @@ enum UnneededSynthesizedInitializerRuleExamples {
                         self.bar = bar
                     }
                 }
+                """),
+        Example("""
+                struct Foo {
+                    init() {
+                        print("perform side effect")
+                    }
+                }
+                """),
+        Example("""
+                struct Foo {
+                    var bar: Int = 0
+
+                    init(bar: Int = 0) {
+                        self.bar = bar
+                        print("perform side effect")
+                    }
+                }
                 """)
     ]
 

